src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -654,6 +654,11 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
         // the array is always zero-initialized.
         internal static T[] AllocateUninitializedArray<T>(int length)
         {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
```

Comment:
Should this be checked only if we are going to call AllocateNewArray? I.e. after the check for min/small size?

Reply:
`IsReferenceOrContainsReferences` is a JIT intrinsic. I expect the rest of the method be short-circuited at JIT time if this is true. Although I did not check if that really happens.

Reply:
`IsReferenceOrContainsReferences` is elided when `T` is a valuetype. When it is a reference type we make a method call to something that returns true. 

I wonder why `IsReferenceOrContainsReferences` does not act as a constant. @jkotas - is that expected?

---
In reply to: [286784972](https://github.com/dotnet/coreclr/pull/24504#discussion_r286784972) [](ancestors = 286784972)

Reply:
That was tiered jitting. 
With tiered JIT disabled reference typesare passed to `T[]` :

```
--- E:\CoreClr2\coreclr\src\System.Private.CoreLib\src\System\GC.cs ------------
            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
00007FF8695FF660  push        rsi  
00007FF8695FF661  sub         rsp,30h  
00007FF8695FF665  xor         eax,eax  
00007FF8695FF667  mov         qword ptr [rsp+20h],rax  
00007FF8695FF66C  mov         qword ptr [rsp+28h],rcx  
00007FF8695FF671  mov         esi,edx  
            {
                return new T[length];
00007FF8695FF673  call        qword ptr [7FF869221A68h]  
00007FF8695FF679  mov         rcx,rax  
00007FF8695FF67C  movsxd      rdx,esi  
00007FF8695FF67F  call        qword ptr [7FF869221098h]  
00007FF8695FF685  nop  
00007FF8695FF686  add         rsp,30h  
00007FF8695FF68A  pop         rsi  
00007FF8695FF68B  ret 
```


---
In reply to: [287150930](https://github.com/dotnet/coreclr/pull/24504#discussion_r287150930) [](ancestors = 287150930,286784972)